### ISSUE TITLE
Question on Endian-ness for Union

### ISSUE THREAD

**Issue by jeffnyman** [craigthomas/Chip8C#2]
It's a little unclear to me what the right way is in terms of the endian-ness logic. Code from the article (http://craigthomas.ca/blog/2014/06/21/writing-a-chip-8-emulator-part-1/) for this code shows the following:

```c
struct {
      #ifdef WORDS_LITTLE_ENDIAN
         byte high, low;
      #else
         byte low, high;
      #endif
} BYTE;
```

Note the WORDS_LITTLE_ENDIAN. The code here in the repo shows:

```c
struct {
      #ifdef WORDS_BIGENDIAN
         byte high, low;
      #else
         byte low, high;
      #endif
} BYTE;
```

Note the WORDS_BIGENDIAN. But the statements themselves -- within the directives -- are not reversed. A statement from the article is:

_"Notice also that the definition of big- and little-endian appear to be backwards - this is because I’m developing on an x86-based system, which is little-endian."_

So ... which is the backwards one? Is the implementation here in the repo correct?

**Comment by craigthomas** [comment]
Sorry for not seeing this issue sooner - for some reason I didn't get a notification for this. The code in the blog post is incorrect, and my description in the blog post is incorrect as well. The code in the repo is correct - I'm sorry for the confusion it's caused. 

As I said on the blog post, I'm developing on an x86 based system, which is indeed little-endian. This means that 16-bit values end up being stored with the MSB in the last memory position. The CHIP 8 is a big-endian architecture, which means the MSB would be stored in the first memory position. For example, on a little-endian machine the 16-bit value `$FFEE` would be stored as `$EEFF` in memory. If we made the union struct just:

```C
struct {
    byte high, low;
} BYTE;
```
then if we run the emulator on a little-endian platform, that value read back would have `high` as `$EE` and `low` as `$FF`, since the structure in memory is nothing more than a 16-bit value which would follow the byte ordering of the architecture. By adding the `#ifdef` in with no environmental pre-processor flags defined, the structure is set to correctly resolve this byte-order discrepancy, and would resolve `low` as `$EE` and `high` as `$FF`. The purpose of the `#ifdef WORDS_BIGENDIAN` would be if you were planning to run the emulator on a big-endian architecture - you would want to supply that in the `CFLAGS` variable as `make CFLAGS=-DWORDS_BIGENDIAN`. Since the byte-ordering of the CHIP 8 would then match the byte-ordering of the host platform, no translation is needed when accessing the memory structure, since the 16-bit value of `$FFEE` is already stored as `$FFEE` in memory.

My comment in the blog post about the definition being backwards makes no sense as far as I can tell. I changed the `#ifdef` variable name for the purposes of the code snippet (the reason why totally eludes me now - it doesn't make any sense why I did that). The code repo is correct. I'll correct the blog post.

Does that answer your question?
